boot/freeldr/uefi/uefivid.c

```diff
@@ -0,0 +1,223 @@
+    TRACE("Framebuffer ScreenWidth       : %d\n", framebufferData.ScreenWidth);
+    TRACE("Framebuffer ScreenHeight      : %d\n", framebufferData.ScreenHeight);
+    TRACE("Framebuffer PixelsPerScanLine : %d\n", framebufferData.PixelsPerScanLine);
+    TRACE("Framebuffer PixelFormat       : %d\n", framebufferData.PixelFormat);
+}
+
+static ULONG
+UefiVideoAttrToSingleColor(UCHAR Attr)
+{
+    UCHAR Intensity;
+    Intensity = (0 == (Attr & 0x08) ? 127 : 255);
+
+    return 0xff000000 |
+           (0 == (Attr & 0x04) ? 0 : (Intensity << 16)) |
+           (0 == (Attr & 0x02) ? 0 : (Intensity << 8)) |
+           (0 == (Attr & 0x01) ? 0 : Intensity);
+}
+
+static VOID
+UefiVideoAttrToColors(UCHAR Attr, ULONG *FgColor, ULONG *BgColor)
+{
+    *FgColor = UefiVideoAttrToSingleColor(Attr & 0xf);
+    *BgColor = UefiVideoAttrToSingleColor((Attr >> 4) & 0xf);
+}
+
+VOID
+UefiVideoClearScreen(UCHAR Attr)
+{
+    ULONG FgColor, BgColor;
+    UINT32 x, y;
+
+    UefiVideoAttrToColors(Attr, &FgColor, &BgColor);
+    for (y = 0; y < framebufferData.ScreenHeight; ++y)
+    {
+        for (x = 0; x < framebufferData.ScreenWidth; ++x)
+        {
+            *((UINT32*)(framebufferData.BaseAddress + 4 * framebufferData.PixelsPerScanLine * (y) + 4 * (x))) = BgColor;
+        }
+    }
+}
+
+#include "../freeldr/arch/vgafont.c"
+
+VOID
+UefiVideoOutputChar(UCHAR Char, unsigned X, unsigned Y, ULONG FgColor, ULONG BgColor)
+{
+    PUCHAR FontPtr;
+    PULONG Pixel;
+    UCHAR Mask;
+    unsigned Line;
+    unsigned Col;
+    ULONG Delta;
+    Delta = (framebufferData.PixelsPerScanLine * 4 + 3) & ~ 0x3;
+    FontPtr = BitmapFont8x16 + Char * 16;
+    Pixel = (PULONG) ((char *) framebufferData.BaseAddress +
+            (Y * CHAR_HEIGHT + TOP_BOTTOM_LINES) *  Delta + X * CHAR_WIDTH * 4);
+
+    for (Line = 0; Line < CHAR_HEIGHT; Line++)
+    {
+        Mask = 0x80;
+        for (Col = 0; Col < CHAR_WIDTH; Col++)
+        {
+            Pixel[Col] = (0 != (FontPtr[Line] & Mask) ? FgColor : BgColor);
+            Mask = Mask >> 1;
+        }
+        Pixel = (PULONG) ((char *) Pixel + Delta);
+    }
+}
+
+VOID
+UefiVideoPutChar(int Ch, UCHAR Attr, unsigned X, unsigned Y)
+{
+    ULONG FgColor = 0;
+    ULONG BgColor = 0;
+    if (Ch != 0)
+    {
+        UefiVideoAttrToColors(Attr, &FgColor, &BgColor);
+        UefiVideoOutputChar(Ch, X, Y, FgColor, BgColor);
+    }
+}
+
+VOID
+UefiVideoGetDisplaySize(PULONG Width, PULONG Height, PULONG Depth)
+{
+    *Width =  framebufferData.ScreenWidth / CHAR_WIDTH;
+    *Height = (framebufferData.ScreenHeight - 2) / CHAR_HEIGHT;
+    *Depth =  0;
+}
+
+VIDEODISPLAYMODE
+UefiVideoSetDisplayMode(char *DisplayMode, BOOLEAN Init)
+{
+    /* We only have one mode, semi-text */
+    return VideoTextMode;
+}
+
+ULONG
+UefiVideoGetBufferSize(VOID)
+{
+    return ((framebufferData.ScreenHeight - 2) / CHAR_HEIGHT * (framebufferData.ScreenWidth / CHAR_WIDTH) * 2);
```

Comment:
Maybe "ScreenHeight - 2 * TOP_BOTTOM_LINES" (regarding the -2)? This is what was done in XBOX.
Also, UefiVideoScrollUp uses a similar construct.